src/storage/src/source/mod.rs

```diff
@@ -952,162 +930,191 @@ where
-            let source_reader = match &mut source_reader {
-                Some(source_reader) => source_reader,
-                None => {
-                    return SourceStatus::Done;
+                    return;
                 }
             };
 
-            // Check that we have a valid list of timestamp bindings.
-            let timestamp_histories = match &mut timestamp_histories {
-                Some(histories) => histories,
-                None => {
-                    error!("Source missing list of timestamp bindings");
-                    return SourceStatus::Done;
+            tokio::pin!(source_stream);
+
+            let mut timestamp_interval = tokio::time::interval(timestamp_frequency);
+            let mut pending_messages = vec![];
+            loop {
+                tokio::select! {
+                    // N.B. This branch is cancel-safe because `next` only borrows the underlying stream.
+                    item = source_stream.next() => {
+                        match item {
+                            Some(Ok(message)) => pending_messages.push(Ok(message)),
+                            // TODO: make errors definite
+                            Some(Err(e)) => pending_messages.push(Err(e)),
+                            None => {},
+                        }
+                    }
+                    // It's time to timestamp a batch
+                    _ = timestamp_interval.tick() => {
+                        let mut max_offsets = HashMap::new();
+                        for message in pending_messages.iter().filter_map(|m| m.as_ref().ok()) {
+                            let entry = max_offsets.entry(message.partition.clone()).or_default();
+                            *entry = std::cmp::max(*entry, message.offset);
+                        }
+                        let (bindings, progress) = match timestamper.timestamp_offsets(max_offsets).await {
+                            Ok(Some((bindings, progress))) => (bindings, progress),
+                            Ok(None) => {
+                                // Timestamper shutdown.  Should exit
+                                break;
+                            },
+                            Err(e) => {
+                                error!("Error timestamping offsets: {}", e);
+                                return;
+                            }
+                        };
+
+                        for msg in pending_messages.drain(..) {
+                            match msg{
+                                Ok(message) => {
+                                    let ts = bindings.get(&message.partition).expect("timestamper didn't return partition").0;
+                                    yield Event::Message(ts, Ok(message));
+                                },
+                                Err(e) => {
+                                    // TODO: make errors definite
+                                    yield Event::Message(0, Err(e));
```

Comment:
Setting the timestamp to `0` will cause timely to panic when you attempt to do `cap.delayed(&ts)` and `cap` is greater than zero